Manage Purchase: Expand the notices that inform about available products for renewal

DavidRothstein opened the pull request:
### Changes proposed in this Pull Request

Following up on https://github.com/Automattic/wp-calypso/pull/42076, which introduced a new feature (behind a feature flag) to show information on the manage purchase pages about other purchases on the same site in some cases where they are expiring soon, this pull requests adds more comprehensive notifications (including for purchases that are auto-renewing soon also), and ensures that the message always shows pertinent information about the current purchase too, when appropriate.

The basic goal is that when viewing one purchase, users always have a way of getting information about (and taking action on) other purchases on the same site that expire or renew soon.

See also internal discussion here: pbOQVh-bC-p2

The pull request handles many cases (too many to list all here) but here are a few new ones to look at.

**Multiple subscriptions on the site are auto-renewing soon:**

![others-renew-soon](https://user-images.githubusercontent.com/235183/82594846-b5fc2880-9b72-11ea-8a16-f9c2618ebffb.png)

**Multiple subscriptions on the site are expiring soon (this was already there, but this pull request makes a small change to the wording to match the above):**

![all-expiring](https://user-images.githubusercontent.com/235183/82594935-d1673380-9b72-11ea-930a-3952922aa164.png)

**The current subscription is auto-renewing soon, but other subscriptions are expiring soon:**

![others-are-expiring](https://user-images.githubusercontent.com/235183/82594956-da580500-9b72-11ea-9e6b-6ce9e879c151.png)

**Everything is auto-renewing, but the current subscription's credit card is expired (since the message refers to multiple subscriptions now, the "Update all" action takes you to https://wordpress.com/me/purchases/add-credit-card for simplicity, to update the credit card for all the subscriptions you own at once):**

![expiring-card](https://user-images.githubusercontent.com/235183/82594978-e17f1300-9b72-11ea-84e5-f3a649a64127.png)

**Since we're showing messages about other subscriptions everywhere, domain mapping subscriptions need to get these kinds of messages too (for consistency), even though they are included in the site's plan and therefore don't normally get messages of their own.  Here is an example of what that looks like:**

![domain-mapping-other-upgrades](https://user-images.githubusercontent.com/235183/82595000-e774f400-9b72-11ea-91eb-7f0ca1d6fa00.png)

**Also for consistency, the above needs to apply even when there aren't multiple subscriptions on the site expiring/renewing soon (for example, a site that only has a plan and domain mapping):**

![domain-mapping-no-other-upgrades](https://user-images.githubusercontent.com/235183/82595032-f360b600-9b72-11ea-9712-208854208c2d.png)

(Note there is an inconsistency between the above two screenshots where the second one doesn't have a "Renew" link.  But as described in the pull request comments, we won't be solving that here.)

### Testing instructions

**Existing functionality:**

The most important thing to test is probably to ensure that existing functionality is unchanged for the time being -- without the `upgrades/upcoming-renewals-notices` flag enabled, there should be no meaningful changes to the Manage Purchase behavior.

The one small exception to that is that I standardized the code that re-styles certain kinds of error messages on these pages as information messages in the case of recently-purchased monthly subscriptions (originally added in https://github.com/Automattic/wp-calypso/pull/23017) to actually apply to monthly subscriptions only.  Without that fix, testing that the red error message appears for expiring subscriptions was a pain -- because it meant that you couldn't just create an annual subscription and then change its expiration date to something soon via the WordPress.com backend (you'd have had to always fake an earlier purchase date too).

**New functionality:**

There's a lot to test here, and I haven't tested all the combinations myself yet (I want to eventually) but to test the basics you might want to set up:
- A WordPress.com site with a plan, domain mapping, a couple domain registrations, a theme, and another purchase.
- A Jetpack site with a plan and an add-on (such as Jetpack Search).

Then, play around with changing their auto-renew settings and expiration dates and make sure the messages you see when viewing each subscription's Manage Purchase page makes sense, and that the message always gives you a way to quickly view the other upgrades that it is talking about and easily change their status in whatever way you want.

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -303,84 +387,406 @@ class PurchaseNotice extends Component {
+			if ( isExpired( currentPurchase ) ) {
+				if ( isDomainRegistration( currentPurchase ) ) {
+					noticeText = translate(
+						'Your %(purchaseName)s domain expired %(expiry)s and will be removed soon unless you take action. {{link}}Other upgrades{{/link}} on this site are scheduled to renew soon.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s domain expired %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -303,84 +387,406 @@ class PurchaseNotice extends Component {
+						);
+					} else {
+						noticeText = translate(
+							'Your %(purchaseName)s plan expired %(expiry)s and will be removed soon unless you take action. {{link}}Other upgrades{{/link}} on this site are scheduled to renew soon.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s plan expired %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -303,84 +387,406 @@ class PurchaseNotice extends Component {
+					}
+				} else {
+					noticeText = translate(
+						'Your %(purchaseName)s subscription expired %(expiry)s and will be removed soon unless you take action. {{link}}Other upgrades{{/link}} on this site are scheduled to renew soon.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s subscription expired %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -303,84 +387,406 @@ class PurchaseNotice extends Component {
+					);
+				} else {
+					noticeText = translate(
+						'Your %(purchaseName)s plan will expire %(expiry)s unless you take action. {{link}}Other upgrades{{/link}} on this site are scheduled to renew soon.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s plan will expire %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 6*

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -303,84 +387,406 @@ class PurchaseNotice extends Component {
+				}
+			} else {
+				noticeText = translate(
+					'Your %(purchaseName)s subscription will expire %(expiry)s unless you take action. {{link}}Other upgrades{{/link}} on this site are scheduled to renew soon.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Your %(purchaseName)s subscription will expire %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 6*

fditrapani commented:
The screenshots look good. One small suggestion for the copy, that can be handled separately because it doesn't look like you added it here, in the case with `Multiple subscriptions on the site are auto-renewing soon`. 

Can we change it from:

Other upgrades on this site are also scheduled to renew soon.

to:

You have other upgrades on this site that are scheduled to renew soon.

DavidRothstein commented:
>Can we change it from:
>
>Other upgrades on this site are also scheduled to renew soon.
>
>to:
>
>You have other upgrades on this site that are scheduled to renew soon.

Yeah, we went with similar wording in some other places (regarding expiring subscriptions) but I think it actually makes sense to use the "You have other upgrades that" wording everywhere.  It is more clear, since it avoids us implying that all subscriptions on their site are necessarily in the mentioned situation.

I've made the above change, and also changed the "other upgrades" wording everywhere else to use a similar pattern also.

**I've updated the screenshots in this pull request to reflect the above, and also to reflect one more change (based on this comment from my original pull request writeup):**

>[when viewing a domain mapping associated with an expiring plan]
>
>(I'm realizing as I write this that for consistency, the above example should maybe get a "Renew Now" link also -- I'll try to add that later.)

It actually turns out this would be a bit involved to do since we need to renew the plan (not the domain mapping that you're looking at) in that case, and the code isn't set up to easily do it.  And as a practical matter, there's another bug (unrelated to this pull request) that seems to suppress those "Renew Now" links pretty often anyway.

So as an alternative for now, I've just made the name of the plan into a link that takes you to the manage purchase page for the plan, and therefore makes it easy for you to go there and renew the plan from there.  I think that's consistent with what we're doing elsewhere.  I also added a code comment explaining this.

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+			if ( isExpired( currentPurchase ) ) {
+				if ( isDomainRegistration( currentPurchase ) ) {
+					noticeText = translate(
+						'Your %(purchaseName)s domain expired %(expiry)s, and you have {{link}}other upgrades{{/link}} on this site that will also be removed soon unless you take action.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s domain expired %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+						);
+					} else {
+						noticeText = translate(
+							'Your %(purchaseName)s plan expired %(expiry)s, and you have {{link}}other upgrades{{/link}} on this site that will also be removed soon unless you take action.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s plan expired %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+					}
+				} else {
+					noticeText = translate(
+						'Your %(purchaseName)s subscription expired %(expiry)s, and you have {{link}}other upgrades{{/link}} on this site that will also be removed soon unless you take action.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s subscription expired %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+			} else if ( anotherPurchaseIsCloseToExpiration ) {
+				if ( isDomainRegistration( currentPurchase ) ) {
+					noticeText = translate(
+						'Your %(purchaseName)s domain will expire %(expiry)s, and you have {{link}}other upgrades{{/link}} on this site that will also be removed soon unless you take action.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s domain will expire %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+						);
+					} else {
+						noticeText = translate(
+							'Your %(purchaseName)s plan will expire %(expiry)s, and you have {{link}}other upgrades{{/link}} on this site that will also be removed soon unless you take action.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s plan will expire %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+					}
+				} else {
+					noticeText = translate(
+						'Your %(purchaseName)s subscription will expire %(expiry)s, and you have {{link}}other upgrades{{/link}} on this site that will also be removed soon unless you take action.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Your %(purchaseName)s subscription will expire %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+				}
+			} else if ( isDomainRegistration( currentPurchase ) ) {
+				noticeText = translate(
+					'Your %(purchaseName)s domain will expire %(expiry)s, and you have {{link}}other upgrades{{/link}} on this site that will also be removed unless you take action.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Your %(purchaseName)s domain will expire %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed unless you take action.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+					);
+				} else {
+					noticeText = translate(
+						'Your %(purchaseName)s plan will expire %(expiry)s, and you have {{link}}other upgrades{{/link}} on this site that will also be removed unless you take action.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Your %(purchaseName)s plan will expire %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed unless you take action.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+				}
+			} else {
+				noticeText = translate(
+					'Your %(purchaseName)s subscription will expire %(expiry)s, and you have {{link}}other upgrades{{/link}} on this site that will also be removed unless you take action.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Your %(purchaseName)s subscription will expire %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed unless you take action.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+			if ( isExpired( currentPurchase ) ) {
+				if ( isDomainRegistration( currentPurchase ) ) {
+					noticeText = translate(
+						'Your %(purchaseName)s domain expired %(expiry)s and will be removed soon unless you take action. You also have {{link}}other upgrades{{/link}} on this site that are scheduled to renew soon.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s domain expired %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+						);
+					} else {
+						noticeText = translate(
+							'Your %(purchaseName)s plan expired %(expiry)s and will be removed soon unless you take action. You also have {{link}}other upgrades{{/link}} on this site that are scheduled to renew soon.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s plan expired %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

a8ci18n commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -300,87 +392,410 @@ class PurchaseNotice extends Component {
+					}
+				} else {
+					noticeText = translate(
+						'Your %(purchaseName)s subscription expired %(expiry)s and will be removed soon unless you take action. You also have {{link}}other upgrades{{/link}} on this site that are scheduled to renew soon.',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Your %(purchaseName)s subscription expired %(expiry)s, and {{link}}other upgrades{{/link}} on this site will also be removed soon unless you take action.' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/expand-manage-purchase-notices-for-related-products)

fditrapani commented:
Thanks for the updates @DavidRothstein! They look great.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/3647867

Thank you @DavidRothstein for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/3647867).
